File: go/datastructures/disjoint_set.go
package datastructures

// DisjointSet implements Union-Find with path compression and union by rank.
// Time: O(α(n)) amortized for all operations (nearly constant)
// Space: O(n)
type DisjointSet struct {
	parent []int
	rank   []int
	count  int
}

func NewDisjointSet(size int) *DisjointSet {
	parent := make([]int, size)
	rank := make([]int, size)
	for i := 0; i < size; i++ {
		parent[i] = i
	}
	return &DisjointSet{
		parent: parent,
		rank:   rank,
		count:  size,
	}
}

func (d *DisjointSet) Find(x int) int {
	if d.parent[x] != x {
		d.parent[x] = d.Find(d.parent[x]) // Path compression
	}
	return d.parent[x]
}

func (d *DisjointSet) Union(x, y int) bool {
	rootX := d.Find(x)
	rootY := d.Find(y)

	if rootX == rootY {
		return false
	}

	// Union by rank
	if d.rank[rootX] < d.rank[rootY] {
		d.parent[rootX] = rootY
	} else if d.rank[rootX] > d.rank[rootY] {
		d.parent[rootY] = rootX
	} else {
		d.parent[rootY] = rootX
		d.rank[rootX]++
	}

	d.count--
	return true
}

func (d *DisjointSet) Connected(x, y int) bool {
	return d.Find(x) == d.Find(y)
}

func (d *DisjointSet) Count() int {
	return d.count
}

func (d *DisjointSet) Size() int {
	return len(d.parent)
}

File: scala/src/main/scala/datastructures/DisjointSet.scala
package datastructures

import scala.collection.mutable

/**
 * Disjoint set (Union-Find) with path compression and union by rank.
 * Time: O(α(n)) amortized for all operations (nearly constant)
 * Space: O(n)
 */
class DisjointSet[T]:
  private val parent = mutable.Map.empty[T, T]
  private val rank = mutable.Map.empty[T, Int]

  def makeSet(x: T): this.type =
    if !parent.contains(x) then
      parent(x) = x
      rank(x) = 0
    this

  def find(x: T): Option[T] =
    if !parent.contains(x) then None
    else Some(findRoot(x))

  private def findRoot(x: T): T =
    if parent(x) != x then
      parent(x) = findRoot(parent(x))
    parent(x)

  def union(x: T, y: T): Boolean =
    if !parent.contains(x) || !parent.contains(y) then false
    else
      val rootX = findRoot(x)
      val rootY = findRoot(y)

      if rootX == rootY then false
      else
        val rankX = rank(rootX)
        val rankY = rank(rootY)

        if rankX < rankY then
          parent(rootX) = rootY
        else if rankX > rankY then
          parent(rootY) = rootX
        else
          parent(rootY) = rootX
          rank(rootX) = rankX + 1
        true

  def connected(x: T, y: T): Boolean =
    (for
      rootX <- find(x)
      rootY <- find(y)
    yield rootX == rootY).getOrElse(false)

  def contains(x: T): Boolean = parent.contains(x)

  def setCount: Int =
    parent.keys.count(x => parent(x) == x)

  def setSize(x: T): Int =
    find(x) match
      case None => 0
      case Some(root) => parent.keys.count(elem => find(elem).contains(root))

  def clear(): this.type =
    parent.clear()
    rank.clear()
    this
